### PR TITLE
feat: Add logical border radius implementation

### PR DESCRIPTION
**Pull request by gabrieldonadel**
## Summary

This PR implements logical border-radius as requested on https://github.com/facebook/react-native/issues/34425. This implementation includes the addition of the following style properties 

- `borderEndEndRadius`, equivalent to `borderBottomEndRadius`.
- `borderEndStartRadius`, equivalent to `borderBottomStartRadius`.
- `borderStartEndRadius`, equivalent to `borderTopEndRadius`.
- `borderStartStartRadius`, equivalent to `borderTopStartRadius`.

## Changelog

[GENERAL] [ADDED] - Add logical border-radius implementation
 

## Test Plan

1. Open the RNTester app and navigate to the `RTLExample` page
2. Test the new style properties through the `Logical Border Radii Start/End` section

https://user-images.githubusercontent.com/11707729/206623732-6d542347-93f9-40da-be97-f7dcd5f66ca9.mov

### CODE REVIEW

**Comment by NickGerleman** [comment]
I *think* what controls this on Android for both Paper and Fabric are all of the `setBorderRadius` functions in ReactAndroid bound to `ViewProps.BORDER_RADIUS` and similar.

The layout props are reused between all views/platforms in Fabric, which is why we can get away with an easy native implementation for it. But for props that control style different from positioning/sizing, the implementation gets a bit more involved.

**Comment by gabrieldonadel** [comment]
Thanks for the explanation @NickGerleman! I'll update this PR so we use this other approach

**Comment by gabrieldonadel** [comment]
Hi @NickGerleman I managed to get this working on iOS on both Fabric and Paper, do you mind taking a look at this PR to check if I'm going in the right direction? I am not too familiar with Yoga so I could have implemented something totally off by accident

File: ReactAndroid/src/main/java/com/facebook/react/views/view/ReactViewBackgroundDrawable.java [hunk 1]
```diff
@@ -598,20 +607,30 @@ private void updatePath() {
         bottomEndRadius = bottomRightRadius;
       }
 
-      final float directionAwareTopLeftRadius = isRTL ? topEndRadius : topStartRadius;
-      final float directionAwareTopRightRadius = isRTL ? topStartRadius : topEndRadius;
-      final float directionAwareBottomLeftRadius = isRTL ? bottomEndRadius : bottomStartRadius;
-      final float directionAwareBottomRightRadius = isRTL ? bottomStartRadius : bottomEndRadius;
+      final float logicalTopStartRadius = !YogaConstants.isUndefined(startStartRadius) ? startStartRadius : topStartRadius;
```
Comment (NickGerleman) [inline ReactAndroid/src/main/java/com/facebook/react/views/view/ReactViewBackgroundDrawable.java:610]: Same comment as above re reversing the precedence here.

**Comment by gabrieldonadel** [comment]
@necolas just FYI this is now ready for review

File: ReactCommon/react/renderer/components/view/YogaLayoutableShadowNode.cpp [hunk 2]
```diff
@@ -751,6 +751,26 @@ void YogaLayoutableShadowNode::swapLeftAndRightInViewProps(
...
+  if (props.borderRadii.endEnd.has_value()) {
+    props.borderRadii.bottomEnd = props.borderRadii.endEnd;
+    props.borderRadii.endEnd.reset();
+  }
+
+  if (props.borderRadii.endStart.has_value()) {
+    props.borderRadii.bottomStart = props.borderRadii.endStart;
+    props.borderRadii.endStart.reset();
+  }
+
```
Comment (NickGerleman) [inline ReactCommon/react/renderer/components/view/YogaLayoutableShadowNode.cpp:773]: Looking at the examples below, they are always changing a left-or-right cardinal direction to a flow-relative one. E.g. nothing is done to `borderRadii.topStart`. But the added code seems to be doing the opposite? E.g. changing start/end to a top-bottom direction.

Are you sure the added code here is needed?

File: ReactCommon/react/renderer/components/view/primitives.h [hunk 3]
```diff
@@ -160,12 +160,24 @@ struct CascadedRectangleCorners {
...
+  OptionalT startStart{};
 
   Counterpart resolve(bool isRTL, T defaults) const {
-    const auto topLeading = isRTL ? topEnd : topStart;
-    const auto topTrailing = isRTL ? topStart : topEnd;
-    const auto bottomLeading = isRTL ? bottomEnd : bottomStart;
-    const auto bottomTrailing = isRTL ? bottomStart : bottomEnd;
+    const auto topLeading = isRTL ? startEnd ? startEnd : topEnd
+        : startStart              ? startStart
+                                  : topStart;
```
Comment (NickGerleman) [inline ReactCommon/react/renderer/components/view/primitives.h:171]: Same comment as elsewhere about precedence, but this ternary is a bit hard to read as-is.  Could we avoid the nested ternary (or maybe this is fine when formatted)?

Comment (gabrieldonadel) [reply ReactCommon/react/renderer/components/view/primitives.h:171]: Yeah that logic was pretty confusing, I've just updated it making things a bit easier to read

File: ReactAndroid/src/main/java/com/facebook/react/views/view/ReactViewBackgroundDrawable.java (hunk 1 above)
Comment (gabrieldonadel) [reply ReactAndroid/src/main/java/com/facebook/react/views/view/ReactViewBackgroundDrawable.java]: Sure, updated

File: React/Views/RCTView.m [hunk 4]
```diff
@@ -668,10 +672,10 @@ - (RCTCornerRadii)cornerRadii
   CGFloat bottomRightRadius;
 
   if ([[RCTI18nUtil sharedInstance] doLeftAndRightSwapInRTL]) {
-    const CGFloat topStartRadius = RCTDefaultIfNegativeTo(_borderTopLeftRadius, _borderTopStartRadius);
-    const CGFloat topEndRadius = RCTDefaultIfNegativeTo(_borderTopRightRadius, _borderTopEndRadius);
-    const CGFloat bottomStartRadius = RCTDefaultIfNegativeTo(_borderBottomLeftRadius, _borderBottomStartRadius);
-    const CGFloat bottomEndRadius = RCTDefaultIfNegativeTo(_borderBottomRightRadius, _borderBottomEndRadius);
+    const CGFloat topStartRadius = RCTDefaultIfNegativeTo(_borderTopLeftRadius, _borderStartStartRadius > 0 ? _borderStartStartRadius: _borderTopStartRadius);
+    const CGFloat topEndRadius = RCTDefaultIfNegativeTo(_borderTopRightRadius, _borderStartEndRadius > 0 ? _borderStartEndRadius: _borderTopEndRadius);
+    const CGFloat bottomStartRadius = RCTDefaultIfNegativeTo(_borderBottomLeftRadius, _borderEndStartRadius > 0 ? _borderEndStartRadius: _borderBottomStartRadius);
```
[the remark this answers is not in the document]
Comment (gabrieldonadel) [reply React/Views/RCTView.m]: Makes sense, I've just updated it

File: React/Views/RCTView.m [hunk 5]
```diff
@@ -668,10 +672,10 @@ - (RCTCornerRadii)cornerRadii
   CGFloat bottomRightRadius;
 
   if ([[RCTI18nUtil sharedInstance] doLeftAndRightSwapInRTL]) {
-    const CGFloat topStartRadius = RCTDefaultIfNegativeTo(_borderTopLeftRadius, _borderTopStartRadius);
-    const CGFloat topEndRadius = RCTDefaultIfNegativeTo(_borderTopRightRadius, _borderTopEndRadius);
-    const CGFloat bottomStartRadius = RCTDefaultIfNegativeTo(_borderBottomLeftRadius, _borderBottomStartRadius);
-    const CGFloat bottomEndRadius = RCTDefaultIfNegativeTo(_borderBottomRightRadius, _borderBottomEndRadius);
+    const CGFloat topStartRadius = RCTDefaultIfNegativeTo(_borderTopLeftRadius, _borderStartStartRadius > 0 ? _borderStartStartRadius: _borderTopStartRadius);
```
[the remark this answers is not in the document]
Comment (gabrieldonadel) [reply React/Views/RCTView.m]: Yeah good point, I guess we could check for `> -1` as all `_border**Radius` variables are initialized using `-1`, but I've refactored this a bit to remove the ternaries and now everything is using `RCTDefaultIfNegativeTo`

**Comment by gabrieldonadel** [comment]
Hi @NickGerleman do you mind checking for me what's this Facebook Internal Linter warning about?

**Comment by NickGerleman** [comment]
> Hi @NickGerleman do you mind checking for me what's this Facebook Internal Linter warning about?

It's complaining because there was a CircleCI failure (unrelated to this change). Everything else is passing minus some clang format warnings we can fixup.

**Comment by NickGerleman** [comment]
Nothing on your side, this should be imported soon (have a bit of a queue atm).

**Pull request closed without merging**